boa_engine/src/lib.rs

```diff
@@ -182,85 +182,268 @@ impl JsArgs for [JsValue] {
     }
 }
 
-/// Execute the code using an existing `Context`.
-///
-/// The state of the `Context` is changed, and a string representation of the result is returned.
 #[cfg(test)]
-pub(crate) fn forward<S>(context: &mut Context<'_>, src: &S) -> String
-where
-    S: AsRef<[u8]> + ?Sized,
-{
-    context
-        .eval_script(Source::from_bytes(src))
-        .map_or_else(|e| format!("Uncaught {e}"), |v| v.display().to_string())
+use std::borrow::Cow;
+
+/// A test action executed in a test function.
+#[cfg(test)]
+#[derive(Clone)]
+pub(crate) struct TestAction(Inner);
+
+#[cfg(test)]
+#[derive(Clone)]
+enum Inner {
+    RunHarness,
+    Run {
+        source: Cow<'static, str>,
+    },
+    InspectContext {
+        op: fn(&mut Context<'_>),
+    },
+    Assert {
+        source: Cow<'static, str>,
+    },
+    AssertEq {
+        source: Cow<'static, str>,
+        expected: JsValue,
+    },
+    AssertWithOp {
+        source: Cow<'static, str>,
+        op: fn(JsValue, &mut Context<'_>) -> bool,
+    },
+    AssertOpaqueError {
+        source: Cow<'static, str>,
+        expected: JsValue,
+    },
+    AssertNativeError {
+        source: Cow<'static, str>,
+        kind: builtins::error::ErrorKind,
+        message: &'static str,
+    },
+    AssertContext {
+        op: fn(&mut Context<'_>) -> bool,
+    },
 }
 
-/// Execute the code using an existing Context.
-/// The str is consumed and the state of the Context is changed
-/// Similar to `forward`, except the current value is returned instead of the string
-/// If the interpreter fails parsing an error value is returned instead (error object)
-#[allow(clippy::unit_arg, clippy::drop_copy)]
 #[cfg(test)]
-pub(crate) fn forward_val<T: AsRef<[u8]> + ?Sized>(
-    context: &mut Context<'_>,
-    src: &T,
-) -> JsResult<JsValue> {
-    use boa_profiler::Profiler;
+impl TestAction {
+    /// Evaluates some utility functions used in tests.
+    pub(crate) const fn run_harness() -> Self {
+        Self(Inner::RunHarness)
+    }
 
-    let main_timer = Profiler::global().start_event("Main", "Main");
+    /// Runs `source`, panickinf if the execution throws.
+    pub(crate) fn run(source: impl Into<Cow<'static, str>>) -> Self {
```

Comment:
What's the reason for using `Cow` string type here instead of just the static string?

Reply:
There are some tests that programmatically generate the tested scripts:
https://github.com/boa-dev/boa/blob/6c0482ecb3b741bb3931087732983680dc9184f3/boa_engine/src/tests/mod.rs#L429-L431
The other option was to take a `String` instead of a `&'static str`, but that would make almost all tests slower, so I opted for taking a `Cow` instead.